### PR TITLE
Feature: add range-map module

### PR DESCRIPTION
**Pull request by lichuang**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Feature: add range-map module

## Changelog

- New Feature

## Related Issues

Fixes #issue

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: common/range-map/src/range_key.rs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+where
+    T: Eq,
+    K: Eq,
+{
+    fn eq(&self, other: &RangeKey<T, K>) -> bool {
+        self.range.start == other.range.start
+            && self.range.end == other.range.end
+            && self.key == other.key
+    }
+}
```
Comment (drmingdrmer) [inline common/range-map/src/range_key.rs:55]: `Eq` can be derived. 🤔

File: common/range-map/src/range_key.rs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+impl<T, K> Ord for RangeKey<T, K>
+where
+    T: Ord + std::fmt::Debug + Copy,
+    K: Ord + std::fmt::Debug + Copy,
+{
+    fn cmp(&self, other: &RangeKey<T, K>) -> Ordering {
+        ((self.range.end, self.range.start), self.key)
+            .cmp(&((other.range.end, other.range.start), other.key))
+    }
```
Comment (drmingdrmer) [inline common/range-map/src/range_key.rs:65]: This method deserves a doc comment explaining why the order is defined as such.

The definition of this `Ord` is critical to the correctness of the range-set.

File: common/range-map/src/range_set.rs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+impl<T, K> Default for RangeSet<T, K>
+where
+    T: Ord + Clone + std::fmt::Debug + Copy,
+    K: Ord + Clone + std::fmt::Debug + Copy + Default,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
```
Comment (drmingdrmer) [inline common/range-map/src/range_set.rs:37]: Such a default can be derived. Save you several lines of code :DDD

File: common/range-map/src/range_set.rs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+impl<T: Debug, K: Debug> Debug for RangeSet<T, K>
+where
+    T: Ord + Clone + std::fmt::Debug + Copy,
+    K: Ord + Clone + std::fmt::Debug + Copy + Default,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_set().entries(self.iter()).finish()
+    }
+}
```
Comment (drmingdrmer) [inline common/range-map/src/range_set.rs:67]: Isn't the derived implementation of `Debug` enough?

File: common/range-map/src/range_set.rs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+    pub fn remove(&mut self, range: Range<T>, k: K) {
+        self.set.remove(&RangeKey::new(range, k));
+    }
+
+    pub fn iter(&self) -> Iter<'_, T, K> {
+        Iter {
+            inner: self.set.iter(),
+        }
+    }
```
Comment (drmingdrmer) [inline common/range-map/src/range_set.rs:114]: Is this enough for you need?
```suggestion
    pub fn iter(&self) -> Iter<'_, T, K> {
            self.set.iter()
    }
```

File: common/range-map/src/range_set.rs [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    // 3. `get_by_key_range(5)` return [2,4],[2,6]
+    // Use the default key when construct `RangeKey::key` for search.
+    pub fn get_by_key_range(&self, key: &T) -> Vec<&RangeKey<T, K>> {
+        let key_as_start = RangeKey::new(*key..*key, K::default());
+
+        self.set
+            .range((Bound::Included(key_as_start), Bound::Unbounded))
+            .filter(|e| {
+                e.range.start.cmp(key) != Ordering::Greater
+                    && e.range.end.cmp(key) != Ordering::Less
```
Comment (drmingdrmer) [inline common/range-map/src/range_set.rs:101]: I think `e.range.end >= key` will always hold. right?

File: Cargo.toml [hunk 7]
```diff
@@ -23,6 +23,7 @@ members = [
     "common/meta/raft-store",
     "common/meta/sled-store",
     "common/meta/types",
+    "common/range-map",
```
Comment (drmingdrmer) [inline Cargo.toml:26]: It is renamed to `range-set` right?

**Comment by Xuanwo** [comment]
This module seems not to depend on any crates in databend, how about moving it out as a separate project?

File: common/range-set/src/range_key.rs [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+impl<T, K> PartialEq for RangeKey<T, K>
+where
+    T: Eq,
+    K: Eq,
+{
+    fn eq(&self, other: &RangeKey<T, K>) -> bool {
+        self.range.eq(&other.range) && self.key.eq(&other.key)
+    }
+}
```
Comment (drmingdrmer) [inline common/range-set/src/range_key.rs:53]: Such a default trait impl can be derived too.

File: common/range-set/src/range_set.rs [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    // 1. `get_by_key_range(1)` return [1,5]
+    // 2. `get_by_key_range(2)` return [1,5],[2,4],[2,6]
+    // 3. `get_by_key_range(5)` return [2,4],[2,6]
+    // Use the default key when construct `RangeKey::key` for search.
+    pub fn get_by_key_range(&self, key: &T) -> Vec<&RangeKey<T, K>> {
+        let key_as_start = RangeKey::new(*key..*key, K::default());
+
+        self.set
+            .range((Bound::Included(key_as_start), Bound::Unbounded))
+            .filter(|e| e.range.start.cmp(key) != Ordering::Greater)
```
Comment (drmingdrmer) [inline common/range-set/src/range_set.rs:58]:
```suggestion
            .filter(|e| e.range.start <= key)
```

File: common/range-set/tests/it/range_key_test.rs [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use core::cmp::Ordering;
+use std::collections::BTreeMap;
+use std::collections::BTreeSet;
+
+use common_range_set::RangeKey;
+
+#[cfg(test)]
```
Comment (drmingdrmer) [inline common/range-set/tests/it/range_key_test.rs:21]: Is this required?

**Review by drmingdrmer — COMMENTED** [review]
The term `key` is used abusively in this crate:
It can be `(end, start, some_thing)` AKA, `RangeKey`, or the last value in `RangeKey`.

`RangeKey` in this crate is `a range and a key`, but the name sounds like `range as a key` or `the key of some range`. Especially when there is a `RangeSet`, which means `a set of ranges`.

BTW `get_by_key_range()` is actually a method to get by `key`.

**Comment by lichuang** [comment]
> The term key is used abusively in this crate:
It can be (end, start, some_thing) AKA, RangeKey, or the last value in RangeKey.

RangeKey in this crate is a range and a key, but the name sounds like range as a key or the key of some range. Especially when there is a RangeSet, which means a set of ranges.

BTW get_by_key_range() is actually a method to get by key.



rename `get_by_key_range` to `get_by_point` and add more comments on `RangeKey`

**Comment by lichuang** [comment]
> This module seems not to depend on any crates in databend, how about moving it out as a separate project?

good suggest! But now this mod is too small to be a a separate project, maybe i will do it sometime later to add more function in this module.

**Review by drmingdrmer — APPROVED** [review]
LGTM~

**Pull request merged**